### PR TITLE
Add Preview Mode

### PR DESCRIPTION
**Pull request by josiahsprague**
Allow certain query parameters to trigger a "preview mode" where all resources are optional.

### CODE REVIEW

**Comment by josiahsprague** [comment]
I broke a couple tests, so I need to fix those, and then write a test for this feature.

**Comment by josiahsprague** [comment]
- [ ] Fix broken tests
- [ ] Add no-index meta tag so preview pages don't get indexed
- [ ] Add tests

File: lib/page.js [hunk 1]
```diff
@@ -181,6 +181,7 @@ var Page = function( page_path, options ){
           if (err) {
             var status = err.status || 500;
             var error  = 'Resource error: ' + err.message;
+            var isOptional = context.isPreview || resource.options.optional;
```
Comment (joanniclaborde) [inline lib/page.js]: We usually use underscore casing for variables, and camel casing for methods, in Solidus... (`is_optional`).

File: lib/page.js [hunk 2]
```diff
@@ -298,7 +299,8 @@ var Page = function( page_path, options ){
         scripts: '<script src="/compiled/scripts.js"></script>',
         styles: '<link rel="stylesheet" href="/compiled/styles.css" />'
       },
-      layout: this.getLayout()
+      layout: this.getLayout(),
+      isPreview: req.query.preview?true:false
```
Comment (joanniclaborde) [inline lib/page.js]: The more "standard" way to turn a value into a boolean is `!!req.query.preview`.

**Comment by joanniclaborde** [comment]
Changes look good, but I'm wondering if we should pick a more specific keyword than `preview`, which could happen with real URLs?

**Comment by josiahsprague** [comment]
Do you mean a more specific query string? A preview situation could happen on an existing URL where there are changes to a post that someone wants to preview.

**Comment by joanniclaborde** [comment]
Yeah a different query string than `preview`, just in case that query string actually exists on a real resource.

**Comment by josiahsprague** [comment]
Well the query string is for the Solidus site, the resources don't get that preview string added.

**Comment by joanniclaborde** [comment]
Unless one of them [is dynamic](https://github.com/solidusjs/solidus/wiki/Pages#dynamic-resources) and uses the `preview` query string value. It's an edge case, but it could happen.

**Comment by josiahsprague** [comment]
So, basically if there's a resource with `{preview}` as a dynamic segment, it could conflict? Any suggestions on a better name? At best we're just making a collision less likely, but is it likely enough that we need to worry about it? I grepped all of my local copies of solidus sites and didn't see any resources that use `{preview}` as a dynamic segment.

**Comment by joanniclaborde** [comment]
Right, as I was saying it's an edge case, I guess if it ever happens we can always rename the dynamic segment... Maybe use `preview_page`?

Also, I think we should also ignore preprocessor errors when in preview mode. Most preprocessors are used to transform resources, and the page will likely blow up if the resource is missing.

**Comment by pushred** [comment]
Since we're talking about this as a "mode" how about following the boolean convention?

`isPreview=true`

**Comment by joanniclaborde** [comment]
Or use Yoda casing: `previewIs=true`

**Comment by josiahsprague** [comment]
I originally had `isPreview=true`, but Joannic said the Solidus convention is to use underscore casing for variables, and camel casing for methods.

**Comment by pushred** [comment]
Yeah actually for query strings I think underscores are better. Less concerned about case, more concerned about wording =)

But we really ought to switch this project over to [semistandard][semistandard] sometime soon. I don't know how I ever lived with all this padding! Then again, I've never contributed any code to solidus!

[semistandard]: https://github.com/Flet/semistandard

**Comment by josiahsprague** [comment]
Wait, are we referring to the query string in the URL or the `context.is_preview` property?

**Comment by pushred** [comment]
I'm talking about the query string.

**Comment by josiahsprague** [comment]
Ok, updated that to `isParam` throughout all of these PRs.

**Comment by joanniclaborde** [comment]
In my first comment, I was talking about the `isPreview` variable name, with the wrong casing... :P

**Comment by josiahsprague** [comment]
Yeah, noticed that now. :stuck_out_tongue:

**Comment by josiahsprague** [comment]
Is this good to merge into master and do a release so we don't have to lock package.json files to this branch?

**Comment by joanniclaborde** [comment]
Before merging, and introducing the new preview modo, I suggest:
 - Skip preprocessor errors
 - Rename the `isOptional` variable into `is_optional`
 - Rename the `isPreview` query parameter into `is_preview`
 - Run the tests, they probably don't pass any more since the query parameter changed

**Comment by josiahsprague** [comment]
How's that look?

**Comment by joanniclaborde** [comment]
Looking good, but I'd change the `req.query.isPreview` param to `req.query.is_preview` (for consistency).

**Comment by josiahsprague** [comment]
Oops, sorry, got mixed messages about that. It's `is_preview` everywhere now. Good to go?

**Comment by joanniclaborde** [comment]
:+1: (please squash your commits before merging!)

**Pull request merged**